### ISSUE TITLE
"Use type casting instead" inspection should probably be disabled by default

### ISSUE THREAD

**Issue by Ocramius** [kalessil/phpinspectionsea#1464]
Consider following code snippet:

```php
class Something
{
    public function __toString() : string {
        return 'ohai!';
    }
}

function gimmeSomething() : Something
{
    // ... 
}

echo gimmeSomething()->__toString();
```

This reports `"(string) gimmeSomething() would express the intent better"`.

I've put a lot of thought in this last year, and I think an explicit `->__toString()` call is preferable. Specifically, if `gimmeSomething()` changes to nullable:

```php

function gimmeSomething() : ?Something
{
    // ... 
}
```

Then the expression `(string) echo gimmeSomething()` keeps working (bug), while `echo gimmeSomething()->__toString();` crashes (null pointer exception, expected).

I've documented this in more detail at https://github.com/ShittySoft/symfony-live-berlin-2018-doctrine-tutorial/pull/3#issuecomment-460441229, but using a `(string)` cast is most probably not to be endorsed.

I'm not sure if this is related to the discussion above (or if it is a core IDE bug), but the following doesn't seem to report a NPE either:

```php
<?php

class Something
{
    public function __toString() : string {
        return 'ohai!';
    }
}

function gimmeSomething() : ?Something
{
    return new Something();
}

// NPE here - not being reported
echo gimmeSomething()->__toString();
```

**Comment by kalessil** [comment]
Done: made this configurable, opted-out by default.

**Comment by Ocramius** [comment]
Thank you 💃
